libraries/shared/watcher/storage_watcher.go

```diff
@@ -79,43 +106,78 @@ func (storageWatcher StorageWatcher) getTransformer(diff utils.StorageDiff) (tra
 func (storageWatcher StorageWatcher) processRow(diff utils.StorageDiff) {
-	storageTransformer, ok := storageWatcher.getTransformer(diff)
-	if !ok {
-		logrus.Debug("ignoring a diff from an unwatched contract")
+	storageTransformer, isTransformerWatchingAddress := storageWatcher.getTransformer(diff)
+	if !isTransformerWatchingAddress {
+		logrus.Debug("ignoring diff from an unwatched contract")
 		return
 	}
+
+	headerID, err := storageWatcher.getHeaderID(diff)
+	if err != nil {
+		logrus.Infof("error getting header for diff: %s", err.Error())
+		storageWatcher.queueDiff(diff)
+		return
+	}
+	diff.HeaderID = headerID
+
 	executeErr := storageTransformer.Execute(diff)
 	if executeErr != nil {
-		logrus.Warn(fmt.Sprintf("error executing storage transformer: %s", executeErr))
-		queueErr := storageWatcher.Queue.Add(diff)
-		if queueErr != nil {
-			logrus.Warn(fmt.Sprintf("error queueing storage diff: %s", queueErr))
-		}
+		logrus.Infof("error executing storage transformer: %s", executeErr.Error())
+		storageWatcher.queueDiff(diff)
 	}
 }
 
 func (storageWatcher StorageWatcher) processQueue() {
 	diffs, fetchErr := storageWatcher.Queue.GetAll()
 	if fetchErr != nil {
-		logrus.Warn(fmt.Sprintf("error getting queued storage: %s", fetchErr))
+		logrus.Infof("error getting queued storage: %s", fetchErr.Error())
 	}
+
 	for _, diff := range diffs {
-		storageTransformer, ok := storageWatcher.getTransformer(diff)
-		if !ok {
-			// delete diff from queue if address no longer watched
+		headerID, getHeaderErr := storageWatcher.getHeaderID(diff)
+		if getHeaderErr != nil {
+			logrus.Infof("error getting header for diff: %s", getHeaderErr.Error())
+			continue
+		}
+		diff.HeaderID = headerID
+
+		storageTransformer, isTransformerWatchingAddress := storageWatcher.getTransformer(diff)
+		if !isTransformerWatchingAddress {
 			storageWatcher.deleteRow(diff.Id)
 			continue
 		}
+
 		executeErr := storageTransformer.Execute(diff)
-		if executeErr == nil {
-			storageWatcher.deleteRow(diff.Id)
+		if executeErr != nil {
+			logrus.Infof("error executing storage transformer: %s", executeErr.Error())
+			continue
 		}
+
+		storageWatcher.deleteRow(diff.Id)
 	}
 }
 
 func (storageWatcher StorageWatcher) deleteRow(id int) {
 	deleteErr := storageWatcher.Queue.Delete(id)
 	if deleteErr != nil {
-		logrus.Warn(fmt.Sprintf("error deleting persisted diff from queue: %s", deleteErr))
+		logrus.Infof("error deleting persisted diff from queue: %s", deleteErr.Error())
+	}
+}
+
+func (storageWatcher StorageWatcher) queueDiff(diff utils.StorageDiff) {
+	queueErr := storageWatcher.Queue.Add(diff)
+	if queueErr != nil {
+		logrus.Infof("error queueing storage diff: %s", queueErr.Error())
+	}
+}
+
+func (storageWatcher StorageWatcher) getHeaderID(diff utils.StorageDiff) (int64, error) {
+	header, getHeaderErr := storageWatcher.HeaderRepository.GetHeader(int64(diff.BlockHeight))
+	if getHeaderErr != nil {
+		return 0, getHeaderErr
+	}
+	if diff.BlockHash.Hex() != header.Hash {
```

Comment:
Thinking maybe this should convert `header.Hash` to a `common.Hash` and do this comparison with `common.Hash` instead of hex strings so that we don't run into pesky issues around the presence/absence of the `0x` prefix

Reply:
done d73a988